File: packages/python/elizaos/features/advanced_memory/evaluators/reflection.py
from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_evaluator_spec
from elizaos.prompts import REFLECTION_TEMPLATE
from elizaos.types import ActionResult, Evaluator, EvaluatorResult, HandlerOptions, ModelType
from elizaos.utils.xml import parse_key_value_xml

if TYPE_CHECKING:
    from elizaos.types import Content, IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_evaluator_spec("REFLECTION")


async def evaluate_reflection(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> EvaluatorResult:
    if state is None:
        return EvaluatorResult(
            score=50,
            passed=True,
            reason="No state for reflection",
            details={},
        )

    recent_interactions: list[str] = []
    room_id = message.room_id

    if room_id:
        recent_messages = await runtime.get_memories(
            room_id=room_id,
            limit=10,
            order_by="created_at",
            order_direction="desc",
        )

        for msg in recent_messages:
            if msg.content and msg.content.text:
                sender = "Unknown"
                if msg.entity_id:
                    if str(msg.entity_id) == str(runtime.agent_id):
                        sender = runtime.character.name
                    else:
                        entity = await runtime.get_entity(msg.entity_id)
                        if entity and entity.name:
                            sender = entity.name
                recent_interactions.append(f"{sender}: {msg.content.text}")

    if not recent_interactions:
        return EvaluatorResult(
            score=50,
            passed=True,
            reason="No recent interactions to reflect on",
            details={"noInteractions": True},
        )

    interactions_text = "\n".join(recent_interactions)

    template = (
        runtime.character.templates.get("reflectionTemplate")
        if runtime.character.templates and "reflectionTemplate" in runtime.character.templates
        else REFLECTION_TEMPLATE
    )
    prompt = runtime.compose_prompt(state=state, template=template)
    prompt = prompt.replace("{{recentInteractions}}", interactions_text)

    response_text = await runtime.use_model(ModelType.TEXT_LARGE, prompt=prompt)
    parsed_xml = parse_key_value_xml(response_text)

    if parsed_xml is None:
        raise ValueError("Failed to parse reflection response")

    quality_str = str(parsed_xml.get("quality_score", "50"))
    quality_score = max(0, min(100, int(quality_str)))

    thought = str(parsed_xml.get("thought", ""))
    strengths = str(parsed_xml.get("strengths", ""))
    improvements = str(parsed_xml.get("improvements", ""))
    learnings = str(parsed_xml.get("learnings", ""))

    passed = quality_score >= 50

    return EvaluatorResult(
        score=quality_score,
        passed=passed,
        reason=f"Strengths: {strengths}\nImprovements: {improvements}",
        details={
            "thought": thought,
            "strengths": strengths,
            "improvements": improvements,
            "learnings": learnings,
            "interactionCount": len(recent_interactions),
        },
    )


async def validate_reflection(
    runtime: IAgentRuntime,
    message: Memory,
    _state: State | None = None,
) -> bool:
    return True


async def _reflection_handler(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
    options: HandlerOptions | None = None,
    callback: Callable[[Content], Awaitable[list[Memory]]] | None = None,
    responses: list[Memory] | None = None,
) -> ActionResult | None:
    """Wrapper handler that matches the expected signature."""
    _ = options, callback, responses  # Unused parameters
    result = await evaluate_reflection(runtime, message, state)
    # Return ActionResult - evaluators don't typically return action results
    return ActionResult(text=result.reason, success=result.passed, values={}, data={})


reflection_evaluator = Evaluator(
    name=str(_spec["name"]),
    description=str(_spec["description"]),
    similes=list(_spec.get("similes", [])) if _spec.get("similes") else [],
    validate=validate_reflection,
    handler=_reflection_handler,
    always_run=bool(_spec.get("alwaysRun", False)),
    examples=[],
)

File: packages/rust/src/features/basic_capabilities/bootstrap/evaluators/reflection.rs
//! REFLECTION evaluator implementation.

use async_trait::async_trait;
use once_cell::sync::Lazy;

use crate::error::{PluginError, PluginResult};
use crate::generated::spec_helpers::require_evaluator_spec;
use crate::prompts::REFLECTION_TEMPLATE;
use crate::runtime::{IAgentRuntime, ModelParams};
use crate::types::{EvaluatorResult, Memory, ModelType, State};
use crate::xml::parse_key_value_xml;

use super::Evaluator;

// Get text content from centralized specs
static SPEC: Lazy<&'static crate::generated::spec_helpers::EvaluatorDoc> =
    Lazy::new(|| require_evaluator_spec("REFLECTION"));

/// Evaluator for reflection on behavior.
pub struct ReflectionEvaluator;

#[async_trait]
impl Evaluator for ReflectionEvaluator {
    fn name(&self) -> &'static str {
        &SPEC.name
    }

    fn description(&self) -> &'static str {
        &SPEC.description
    }

    async fn validate(&self, _runtime: &dyn IAgentRuntime, _message: &Memory) -> bool {
        true
    }

    async fn evaluate(
        &self,
        runtime: &dyn IAgentRuntime,
        message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<EvaluatorResult> {
        let room_id = match message.room_id {
            Some(id) => id,
            None => {
                return Ok(EvaluatorResult::pass(50, "No room for reflection")
                    .with_detail("noInteractions", true));
            }
        };

        // Get recent messages for reflection
        let recent_messages = runtime
            .get_memories(Some(room_id), None, None, 10)
            .await
            .unwrap_or_default();

        if recent_messages.is_empty() {
            return Ok(
                EvaluatorResult::pass(50, "No recent interactions to reflect on")
                    .with_detail("noInteractions", true),
            );
        }

        // Format interactions
        let mut interactions = Vec::new();
        for msg in recent_messages.iter().rev() {
            if msg.content.text.is_empty() {
                continue;
            }

            let sender = if let Some(entity_id) = msg.entity_id {
                if entity_id == runtime.agent_id() {
                    runtime.character().name.clone()
                } else if let Ok(Some(entity)) = runtime.get_entity(entity_id).await {
                    entity.name.unwrap_or_else(|| "Unknown".to_string())
                } else {
                    "Unknown".to_string()
                }
            } else {
                "Unknown".to_string()
            };

            interactions.push(format!("{}: {}", sender, msg.content.text));
        }

        let interactions_text = interactions.join("\n");

        // Get template and compose prompt
        let template = runtime
            .character()
            .templates
            .get("reflectionTemplate")
            .map(|s| s.as_str())
            .unwrap_or(REFLECTION_TEMPLATE);

        let composed_state = runtime.compose_state(message, &["RECENT_MESSAGES"]).await?;

        let prompt = runtime
            .compose_prompt(&composed_state, template)
            .replace("{{recentInteractions}}", &interactions_text);

        // Call the model
        let response = runtime
            .use_model(ModelType::TextLarge, ModelParams::with_prompt(&prompt))
            .await
            .map_err(|e| PluginError::ModelError(e.to_string()))?;

        let response_text = response
            .as_text()
            .ok_or_else(|| PluginError::ModelError("Expected text response".to_string()))?;

        // Parse XML response
        let parsed = parse_key_value_xml(response_text).ok_or_else(|| {
            PluginError::XmlParse("Failed to parse reflection response".to_string())
        })?;

        let thought = parsed.get("thought").cloned().unwrap_or_default();
        let quality_str = parsed
            .get("quality_score")
            .cloned()
            .unwrap_or_else(|| "50".to_string());
        let strengths = parsed.get("strengths").cloned().unwrap_or_default();
        let improvements = parsed.get("improvements").cloned().unwrap_or_default();
        let learnings = parsed.get("learnings").cloned().unwrap_or_default();

        let quality_score: u8 = quality_str.parse().unwrap_or(50).min(100);

        let passed = quality_score >= 50;
        let reason = format!("Strengths: {}\nImprovements: {}", strengths, improvements);

        let result = if passed {
            EvaluatorResult::pass(quality_score, &reason)
        } else {
            EvaluatorResult::fail(quality_score, &reason)
        };

        Ok(result
            .with_detail("thought", thought)
            .with_detail("strengths", strengths)
            .with_detail("improvements", improvements)
            .with_detail("learnings", learnings)
            .with_detail("interactionCount", interactions.len() as i64))
    }
}
